Autogen intervals issue5

nandan75 opened the pull request:
Hi @georgiesamaha 

I have pushed the changes related to issue  #5 
The user can create genomic interval two ways :
- (A) Optimised number of genomic intervals based on genome size - The only additional input required is a .dict file
- (B) Specific number of genomic intervals based on a variable `number_of_intervals`

For testing please use the script 
- [run_pipeline_on_gadi_script.sh](https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/blob/autogen_intervals_issue5/run_pipeline_on_gadi_script.sh)
- You will need to hash (deactivate) the variable `number_of_intervals=50` to toggle between options (A) and (B)

I have tested the feature and updated the Readme only for the issue #5 . 
A few checks (say for missing dict file for option A etc are pending)
If you are busy I will proceed to merge it to the main and work with the other requirements. 

Cheers,

Nandan

Commit:
Tested Autogenerate number of intervals both specific number and genome size based  #5

Commit:
Updated Readme.md  #5

georgiesamaha commented on README.md:
```diff
@@ -73,27 +73,37 @@ This pipeline uses the following tools for generating specific index files.
 
-***Note***: You must specify the full path for the reference fasta, even if it is in your working directory.
+***Note***:  
+You must specify the full path for the reference fasta, even if it is in your working directory.
```
Not really sure what this is referring to. You explain how to provide reference file with `--ref` in step 4 so this can probably be removed here.

georgiesamaha commented on README.md:
```diff
@@ -73,27 +73,37 @@ This pipeline uses the following tools for generating specific index files.
+Interval lists define subsets of genomic regions in the genome. It is advisable to split the genome into genomic-intervals to parallelize execution and/or exclude regions that have bad or uninformative data. Please refer to this [link](https://gatk.broadinstitute.org/hc/en-us/articles/360035531852-Intervals-and-interval-lists) for more details. 
+
+Documented below are the two different methods by which you can generate the interval files for Mutect2 when you run the pipeline using the nextflow command.  
+**(a) Generate optimised number of intervals based on your genome size**
```
```suggestion
**Generate optimised number of intervals based on your genome size**
```

georgiesamaha commented on README.md:
```diff
@@ -73,27 +73,37 @@ This pipeline uses the following tools for generating specific index files.
+Documented below are the two different methods by which you can generate the interval files for Mutect2 when you run the pipeline using the nextflow command.  
+**(a) Generate optimised number of intervals based on your genome size**
+  - Provide the [picard](https://gatk.broadinstitute.org/hc/en-us/articles/360037593331-CreateSequenceDictionary-Picard-).dict file in addition to the Reference FASTA file in the command.
+  - Please **do not** provide the `--number_of_intervals` value in the command.
```
Given this section is about preparing input data you don't want to confuse people by providing unnecessary details at this stage. Please remove this line or at least move to the relevant section of instructions for running the pipeline.

georgiesamaha commented on README.md:
```diff
@@ -73,27 +73,37 @@ This pipeline uses the following tools for generating specific index files.
+  - Please **do not** provide the `--number_of_intervals` value in the command.
+  - This generates minimum 100 Mb sized intervals e.g.  32 genomic-interval files for the Human genome hg38.
+
+**(b) Generate a specific number of intervals**  
```
```suggestion
**Generate a specific number of intervals**  
```

georgiesamaha commented on README.md:
```diff
@@ -73,27 +73,37 @@ This pipeline uses the following tools for generating specific index files.
+**(b) Generate a specific number of intervals**  
+Alternatively, based on the computing resources available on the your operating system, you might want to generate a specific number of genomic-intervals.
+  - Please provide an integer value `--number_of_intervals` in the command. 
+  - You do not need to provide the [picard](https://gatk.broadinstitute.org/hc/en-us/articles/360037593331-CreateSequenceDictionary-Picard-).dict file for this option.
```
Instead of saying what users don't need to do, say what they do need to do.

georgiesamaha commented on README.md:
```diff
@@ -73,27 +73,37 @@ This pipeline uses the following tools for generating specific index files.
+
+
+#### Create a Panel of Normals (PoN) for Mutect2  
 The user can create a PoN file using our [pipeline](https://github.com/Sydney-Informatics-Hub/Somatic-ShortV?tab=readme-ov-file#4-create-pon-per-genomic-interval).
```
Given this is current not properly documented just yet, suggest we remove this section and instead point to documentation from GATK about how to make PoN instead.

georgiesamaha commented on README.md:
```diff
@@ -73,27 +73,37 @@ This pipeline uses the following tools for generating specific index files.
 
-#### Common biallelic variant resources for GetPileupSummaries 
+#### Create common biallelic variant resources for GetPileupSummaries 
 The user can create the Common biallelic variant resource files using our [pipeline](https://github.com/Sydney-Informatics-Hub/Somatic-ShortV?tab=readme-ov-file#0-optional-create-common-biallelic-variant-resources)
```
Why hasn't this feature been built into this pipeline if it is a direct porting of SomaticShortV? 

Please create an issue describing this feature to be added and we can address it later, remove this recommendation from the README as the intention behind Nextflowing the pipeline was to reduce the workflow execution steps down to 1. 

If what you've linked here works as a standalone script, please add a copy of it to this repository in the `Scripts` directory and link to that here, mentioining this functionality will be included in the workflow.

nandan75 replied:
- I have now added the script [gatk4_selectvariants.pbs](https://github.com/Sydney-Informatics-Hub/Somatic-ShortV/blob/master/gatk4_selectvariants.pbs) which takes your public resource VCF, and selects common biallelic SNP variants (by default, those with an allele frequency of > 0.05) to **Scripts** folder

- The README.md has been changed accordingly 

- This will be added as a optional module in version 2 of the nextflow pipeline

georgiesamaha commented on README.md:
````diff
@@ -126,7 +136,7 @@ The important features are:
 
 ```
-nextflow run main.nf --input samples.csv --ref reference.fasta  --intervalList_path path_to_intervals --ponvcf pon.vcf.gz
+nextflow run main.nf --input samples.csv --ref reference.fasta --ponvcf pon.vcf.gz
````
This is not accurate. If I run this I get the help message and no error but the workflow doesn't run: 

```bash
NOTE: Nextflow is not tested with Java 1.8.0_392 -- It's recommended the use of version 11 up to 18

N E X T F L O W  ~  version 22.04.3
Launching `main.nf` [deadly_heyrovsky] DSL2 - revision: 28fee1ea87

===================================================================
===================================================================
SOMATIC SHORT V - NF 
===================================================================
===================================================================

Created by the Sydney Informatics Hub, University of Sydney

Documentation   @ https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf

Cite                                    @ https://doi.org/10.48546/workflowhub.workflow.691.1

Log issues    @ https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/issues

All the default parameters are set in `nextflow.config`

=======================================================================================
Workflow run parameters 
=======================================================================================
version           : 1.0
input             : /scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/samples.csv
reference         : /g/data/er01/SIH-HPC-WGS/Reference/hs38DH.fasta
small_exac_common : 
ponvcf            : /scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/pon.vcf.gz
intervalList_path : 
outDir            : results
workDir           : /scratch/er01/gs5517/workflowTesting/Somatic-shortV-nf/work

=======================================================================================

 

Usage:   nextflow run main.nf --input samples.csv --ref reference.fasta --ponvcf pon.vcf.gz

Required Arguments:
  --input                                     Full path and name of sample input file (csv format)
 --ref                                        Full path and name of reference genome (fasta format)
  --ponvcf                        Full path and name of the Panel of Normals (ponvcf) file

Optional Arguments:
  --outDir                        Specify name of results directory. 
  --dict                          Full path and name of reference genome dictionary file (contains metadata and sequence dictionary information)
                                  The above dict file is required when the user chooses to create optimised number of intervals based on genome size

HPC accounting arguments:
  --whoami                    HPC user name (Setonix or Gadi HPC)
  --gadi_account              Project accounting code for NCI Gadi (e.g. aa00)
  ```

nandan75 replied:
I have tried to identify the mandatory parameters and a couple of optional ones.

I have a question which I am unable to get my head around. Can you suggest what could be the problem?

In the run_pipeline_on_gadi_script.sh script , if
-  I remove any of the mandatory parameters from the command , e.g. **--ref ${ref}** from the command below , the if condition in main.nf catches it and displays the help dialog and prevents execution of the pipeline.

nextflow run main.nf \
        --input ${samples} \
        -profile gadi \
        --whoami $(whoami) --gadi_account $PROJECT \
         --dict ${dict} \
	--common_biallelic_variants ${common_biallelic_variants} \
        --ponvcf ${ponvcf} \
        --outDir ${outDir} \
        --number_of_intervals ${number_of_intervals}

However if 
- I inactivate (using #) the parameter
ref=/g/data/er01/SIH-HPC-WGS/Reference/hs38DH.fasta
 but retain  it in the command as below
nextflow run main.nf \
        --input ${samples} \
        -profile gadi \
        --whoami $(whoami) --gadi_account $PROJECT \
        **--ref ${ref}** --dict ${dict} \
	--common_biallelic_variants ${common_biallelic_variants} \
        --ponvcf ${ponvcf} \
        --outDir ${outDir} \
        --number_of_intervals ${number_of_intervals}
        
The value displayed for ref = true and I get an error!

I am concerned that the user can do one of the above (by mistake) and both the options should lead to identical help display message!

Thanks.

georgiesamaha replied:
Bit confused by this, why would anyone hash the `ref=` variable in the bash script if it is a required argument and your instructions document that?   

You've used an if statement to ensure the pipeline doesn't run unless all those required params are provided: https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/blob/0e9f22d1b3bc1b4ff7b43aa1fc5db78afba41ed9/main.nf#L83

The `ref` parameter is defined as false in `nextflow.config`: https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/blob/0e9f22d1b3bc1b4ff7b43aa1fc5db78afba41ed9/nextflow.config#L18

But you're still passing an (albeit empty) value via your run command when using `--ref ${ref}`. To nextflow it might look like `--ref <space>` so that might be why its displaying as `true`.

nandan75 replied:
> Bit confused by this, why would anyone hash the `ref=` variable in the bash script if it is a required argument and your instructions document that?

**By mistake I guess**; So I was thinking of stopping the user .. But I agree. `ref=` variable is a mandatory requirement.



> You've used an if statement to ensure the pipeline doesn't run unless all those required params are provided:
> 
> https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/blob/0e9f22d1b3bc1b4ff7b43aa1fc5db78afba41ed9/main.nf#L83

BTW The above if condition is not a part of the updated branch main.nf at https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/blob/autogen_intervals_issue5/main.nf
This is how it looks like now: 
```
  if ( params.help == true || params.ref == false || params.input == false || params.ponvcf == false || params.common_biallelic_variants == false || params.dict == false)
```

> 
> The `ref` parameter is defined as false in `nextflow.config`:
> 
> https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/blob/0e9f22d1b3bc1b4ff7b43aa1fc5db78afba41ed9/nextflow.config#L18
> 
> But you're still passing an (albeit empty) value via your run command when using `--ref ${ref}`. To nextflow it might look like `--ref <space>` so that might be why its displaying as `true`.
**OK Thanks.**

georgiesamaha replied:
What do you mean stopping the user? How would you do this?

georgiesamaha commented on README.md:
````diff
@@ -126,7 +136,7 @@ The important features are:
+nextflow run main.nf --input samples.csv --ref reference.fasta --ponvcf pon.vcf.gz
         
 ```
 **Note**: Please use the command provided in the script `scripts/run_pipeline_on_gadi_script.sh` on NCI Gadi HPC.
````
You need to explain why and how to apply each parameter including what is required and what is optional. This pipeline is also written for Gadi only at this stage and you're missing any explanation of how to engage the job scheduler (i.e. use gadi account and whoami

nandan75 replied:
Changed

I have now added 
- A basic command
- The basic command to be run specifically on NCI Gadi (with reference to the job schedular paramters)
- An advanced command with a couple of optional parameters

georgiesamaha commented on main.nf:
```diff
@@ -56,16 +57,17 @@ workDir           : ${workflow.workDir}
 def helpMessage() {
     log.info"""
-  Usage:   nextflow run main.nf --input samples.csv --ref reference.fasta  --intervalList_path path_to_intervals --ponvcf pon.vcf.gz
+  Usage:   nextflow run main.nf --input samples.csv --ref reference.fasta --ponvcf pon.vcf.gz
```
See above. This needs to be changed when you've identified what all required arguments are.

georgiesamaha commented on main.nf:
```diff
@@ -56,16 +57,17 @@ workDir           : ${workflow.workDir}
-  Usage:   nextflow run main.nf --input samples.csv --ref reference.fasta  --intervalList_path path_to_intervals --ponvcf pon.vcf.gz
+  Usage:   nextflow run main.nf --input samples.csv --ref reference.fasta --ponvcf pon.vcf.gz
 
   Required Arguments:
```
Lacks intervals

nandan75 replied:
The intervals files are generated on the fly
- Either optimised number based on genome size
- Or an user defined value `number_of-intervals`

So I do not pass them in the command

georgiesamaha commented on main.nf:
```diff
@@ -80,13 +82,16 @@ workflow {
 // provided at runtime
 
-  if ( params.help == true || params.ref == false || params.input == false || params.ponvcf == false || params.small_exac_common == '' || params.intervalList_path == '')
+  if ( params.help == true || params.ref == false || params.input == false || params.ponvcf == false || params.small_exac_common == '' || (params.number_of_intervals == 0 && params.dict == false))
```
intervals and dict path are included as required intervals here which is why I couldn't run the pipeline with your default command above.

nandan75 replied:
Sure .. I have removed it now.

georgiesamaha commented on main.nf:
```diff
@@ -100,9 +105,16 @@ workflow {
+// Create an intervalList based on user specified number_of_intervals
+ createIntervalLists(params.number_of_intervals)
+
+
```
Please tidy formatting and remove all extra new lines

georgiesamaha commented on main.nf:
```diff
@@ -100,9 +105,16 @@ workflow {
+
+
+ intervalList=createIntervalLists.out[2].tokenize(" ") 
+// Run mutect2 on a Tumor/Normal sample-pair
```
Please add a new line between 114-115

georgiesamaha commented on modules/createIntervalLists.nf:
```diff
@@ -10,36 +10,82 @@ input:
+            -scatter-count $number_of_intervals \
+         -O intervals_folder
+
+
```
Please remove extra new line

georgiesamaha commented on modules/createIntervalLists.nf:
```diff
@@ -10,36 +10,82 @@ input:
+
+    number_of_intervals=$(expr $size / $min)
+
+
```
Please remove extra new lines

georgiesamaha commented on modules/createIntervalLists.nf:
```diff
@@ -10,36 +10,82 @@ input:
+    for ((i = 0; i <= (number_of_intervals-1); i++)); do
+        printf "%04d\n" "$i" >> intervals_list.txt
+    done
+
```
Please remove extra new line

georgiesamaha commented on modules/createIntervalLists.nf:
```diff
@@ -10,36 +10,82 @@ input:
+        intervalList+=("$line")
+    done < "intervals_list.txt"
+
+
```
Please remove extra new lines

georgiesamaha commented on modules/createIntervalLists.nf:
```diff
@@ -10,36 +10,82 @@ input:
+            -scatter-count $number_of_intervals \
+         -O intervals_folder
+
+
```
Please remove extra new lines

georgiesamaha commented on modules/createIntervalLists.nf:
```diff
@@ -10,36 +10,82 @@ input:
-        -scatter-count !{number_of_intervals} \
-        -O BQSR_intervals
 
 #-XL !{params.exclude_intervals} \
```
If this is not used, remove it

georgiesamaha commented on modules/mutect2.nf:
```diff
@@ -32,12 +38,16 @@ process mutect2 {
 
 
 
 
```
Please remove extra new lines

georgiesamaha commented on nextflow.config:
```diff
@@ -21,8 +21,8 @@ params {
+  dict                          = false
   
   
   
```
Please remove extra new lines

georgiesamaha commented on run_pipeline_on_gadi_script.sh:
```diff
@@ -0,0 +1,56 @@
+
+export NXF_SINGULARITY_CACHEDIR=/scratch/$PROJECT/$(whoami)/singularity
+
+# Fill in these variables for your run
+samples=/scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/samples.csv
+ponvcf=/scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/pon.vcf.gz
+ref=/g/data/er01/SIH-HPC-WGS/Reference/hs38DH.fasta
+small_exac_common=/g/data/er01/SIH-HPC-WGS/Reference/gatk-best-practices/somatic-hg38/small_exac_common_3.hg38.vcf.gz
+intervalList_path=/scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/interval_files
+outDir=results
+
+
+# Select one of the following options
+	# (A)  
+	# - Generate optimised number of intervals based on your chosen genome and its size
+	# - Please do the following:
+		# - Hash to inactivate the `number_of_intervals` variable below
+		# - You will also need to provide path to reference genome .dict file
+			dict=/g/data/er01/SIH-HPC-WGS/Reference/hs38DH.dict
+
+	# OR 
+
+	# (B) 
+	# Generate an arbitrary number of intervals between 2 and 100 based on your system resources
+		#number_of_intervals=50
```
None of this is necessary and doesn't actually make much sense given your required inputs specified in `main.nf`. You can opt to define a variable if it is being applied otherwise it does not need to be included. Remove this whole section and instead add `dict=` (which you've actually mandated as a required parameter) and `number_of_intervals=` with default values and describe clearly in which situations and how to apply each variable in the relevant section of the README. 

Additionally, a comment like 'generate an arbirtary number of intervals between 2 and 100 based on your system resources' needs to be explained. How would I know what is an appropriate number and what you mean by system resources?

georgiesamaha replied:
```suggestion
# Fill in these variables for your run
samples=/scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/samples.csv
ponvcf=/scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/pon.vcf.gz
ref=/g/data/er01/SIH-HPC-WGS/Reference/hs38DH.fasta
small_exac_common=/g/data/er01/SIH-HPC-WGS/Reference/gatk-best-practices/somatic-hg38/small_exac_common_3.hg38.vcf.gz
intervalList_path=/scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/interval_files
outDir=results
dict=/g/data/er01/SIH-HPC-WGS/Reference/hs38DH.dict
number_of_intervals=50
```

georgiesamaha commented on run_pipeline_on_gadi_script.sh:
```diff
@@ -0,0 +1,56 @@
+	    --small_exac_common ${small_exac_common}\
+        --ponvcf ${ponvcf} \
+        --outDir ${outDir} \
+        --number_of_intervals ${number_of_intervals} 
```
Running this gives: 
```
WARN: Access to undefined parameter `exclude_intervals` -- Initialise it to a default value eg. `params.exclude_intervals = some_value`
```

And `intervalList_path` is not defined in workflow run parameters: 
```
=======================================================================================
Workflow run parameters 
=======================================================================================
version           : 1.0
input             : /scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/samples.csv
reference         : /g/data/er01/SIH-HPC-WGS/Reference/hs38DH.fasta
small_exac_common : /g/data/er01/SIH-HPC-WGS/Reference/gatk-best-practices/somatic-hg38/small_exac_common_3.hg38.vcf.gz
ponvcf            : /scratch/er01/ndes8648/pipeline_work/nextflow/INFRA-83-Somatic-ShortV/test_files_for_Georgie/pon.vcf.gz
intervalList_path : 
outDir            : results
workDir           : /scratch/er01/gs5517/workflowTesting/Somatic-shortV-nf/work

=======================================================================================
```

georgiesamaha reviewed (changes requested):
Hi Nandan, 

Please see to all comments in this review and request my review again once they are complete. I can execute the command you provide in the `run_pipeline_on_gadi_script.sh` however the instructions in your README are incomplete and inaccurate and I cannot follow them to execute the pipeline. 

Additionally, we don't want to point people back to somaticShortV pipeline to set up all required inputs, everything required to run the pipeline should either be provided or described here. For sections lacking functionality in this version of the pipeline, please either make a note in the README that they are optional and functionality will be added and link out to the GATK documentation or add a script to this repo. Please keep your end user in mind when writing instructions for execution.

Commit:
Updated Readme.md  #5, added gatk4_selectvariants.pbs

Commit:
more Updatesd Readme.md and main.nf #5

Commit:
Readme.md typos etc #5

nandan75 commented:
> Hi Nandan,
> 
> Please see to all comments in this review and request my review again once they are complete. I can execute the command you provide in the `run_pipeline_on_gadi_script.sh` however the instructions in your README are incomplete and inaccurate and I cannot follow them to execute the pipeline.
> 
> Additionally, we don't want to point people back to somaticShortV pipeline to set up all required inputs, everything required to run the pipeline should either be provided or described here. For sections lacking functionality in this version of the pipeline, please either make a note in the README that they are optional and functionality will be added and link out to the GATK documentation or add a script to this repo. Please keep your end user in mind when writing instructions for execution.

=====================

Hi @georgiesamaha 

I have now addressed all the points except 1 
I have made changes to the README.md file to make it more user friendly. Please have a look

I have a question- Can you please suggest what could be going wrong here -https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/pull/10#discussion_r1446825512
Thanks

You can test the pipeline as before using

[run_pipeline_on_gadi_script.sh](https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/blob/autogen_intervals_issue5/run_pipeline_on_gadi_script.sh)

georgiesamaha commented on README.md:
```diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
-  - Reference FASTA files must be accompanied by a .fai index file. 
-  - You can download FASTA files from the [Ensembl](https://asia.ensembl.org/info/data/ftp/index.html), [UCSC](https://genome.ucsc.edu/goldenPath/help/ftp.html), or [NCBI](https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/) ftp sites. 
+- Genomic-interval list files for Mutect2
+- Panel of Normals (PoN) file for Mutect2
```
PoN doesnt really qualify as reference material as it is made for each cohort. I suggest including this in the inputs section instead.

georgiesamaha commented on README.md:
```diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
+- Common biallelic variant resources for GetPileupSummaries 
 
-- You can use our [IndexReferenceFasta-nf pipeline](https://github.com/Sydney-Informatics-Hub/IndexReferenceFasta-nf) to generate indexes. 
+You can download FASTA files from the [Ensembl](https://asia.ensembl.org/info/data/ftp/index.html), [UCSC](https://genome.ucsc.edu/goldenPath/help/ftp.html), or [NCBI](https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/) ftp. sites. Reference FASTA files must be accompanied by specific index files. You can use our [IndexReferenceFasta-nf pipeline](https://github.com/Sydney-Informatics-Hub/IndexRefere ...
```
Changes here are required for resolving https://github.com/Sydney-Informatics-Hub/Somatic-shortV-nf/issues/7. Please ensure you work on this section in a new PR once this PR is resolved. Ignoring for now.

georgiesamaha commented on README.md:
````diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
+--ref /path/to/ref.fasta
+```
+
+**Create genomic-interval list files using the pipeline**  
````
Tidy formatting: 

```suggestion
**Create genomic-interval list files using the pipeline**  
  
```

georgiesamaha commented on README.md:
````diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
+When you run the pipeline, you will use the mandatory --ref parameter to specify the location and name of the reference.fasta file:
+```
+--ref /path/to/ref.fasta
+```
````
Tidy formatting, add a header for consistency with other sections below 
```suggestion
**Reference genome indexes**

When you run the pipeline, you will use the mandatory `--ref` parameter to specify the location and name of the reference.fasta file:
\```
--ref /path/to/ref.fasta
\```
```

georgiesamaha commented on README.md:
````diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
+When you run the pipeline, you will use the mandatory --ref  and --dict parameters to specify the location and name of the reference files. The genomic-interval files will be generated by the pipeline based on these two parameters.
+```
+--ref /path/to/ref.fasta --dict /path/to/ref.dict
+```
````
If this doesnt require a user to prepare any input files then it doesn't need to be included in this section.

georgiesamaha commented on README.md:
````diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
+```
 
+***Generate a specific number of intervals (Optional)***   
+You can define a specific number genomic-intervals based on the number of cpus available to parallelize execution. When you run the pipeline, you will use the optional `--number_of_intervals` parameter as an integer value. The genomic-interval files will be generated by the pipeline based on this parameter.
````
Same as above, this section is about preparing input and ref files, not demonstrating how to execute the pipeline

georgiesamaha commented on README.md:
```diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
 - Indexed reference genome in FASTA format. 
-  - Reference FASTA files must be accompanied by a .fai index file. 
-  - You can download FASTA files from the [Ensembl](https://asia.ensembl.org/info/data/ftp/index.html), [UCSC](https://genome.ucsc.edu/goldenPath/help/ftp.html), or [NCBI](https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/) ftp sites. 
+- Genomic-interval list files for Mutect2
```
Not sure why this point is included here if the pipeline creates intervals on the fly (whether it be automated or using a user-specified number of intervals). What input/ref materials need to be prepared for this? If it is only fasta and dict, then this point is irrelevant.

georgiesamaha commented on README.md:
```diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
-#### A Panel of Normals (PoN) for Mutect2  
-The user can create a PoN file using our [pipeline](https://github.com/Sydney-Informatics-Hub/Somatic-ShortV?tab=readme-ov-file#4-create-pon-per-genomic-interval).
+#### Create a Panel of Normals (PoN) for Mutect2  
+The user can create a a panel of normals (PoN) containing germline and artifactual sites for use with Mutect2 using the instructions provided [here](https://gatk.broadinstitute.org/hc/en-us/articles/360037058172-CreateSomaticPanelOfNormals-BETA-). We will include this functionality as an optional module in the next version of the pipeline.
```
```suggestion
You will need to provide a [panel of normals (PoN)](https://gatk.broadinstitute.org/hc/en-us/articles/360037058172-CreateSomaticPanelOfNormals-BETA-) containing germline and artifactual sites for use with Mutect2. This functionality is currently under development and will be included in this pipeline shortly.  
```

georgiesamaha replied:
Ensure the version of GATK documentation you're linking to corresponds with the version of GATK being run by your pipeline

nandan75 replied:
Changed the link to gatk4.4

georgiesamaha commented on README.md:
```diff
@@ -123,25 +133,53 @@ The important features are:
 ### 4. Run the pipeline 
 
-The most basic run command for this pipeline is: 
+The most basic run command for this pipeline with all required paramters looks like this: 
```
```suggestion
The minimal run command for this executing this pipeline is: 
```

georgiesamaha commented on README.md:
```diff
@@ -123,25 +133,53 @@ The important features are:
 By default, this will generate `work` directory, `results` output directory and a `runInfo` run metrics directory inside the results directory. 
 
+
+This pipeline has been tested for NCI Gadi HPC and the paramaters `--whoami` and `--gadi_account` will have to be included as discussed in the section [Infrastructure usage and recommendations](#infrastructure-usage-and-recommendations) .
```
```suggestion

**Running on NCI Gadi HPC**

This pipeline has been optimised for NCI Gadi HPC, instructions for executing on Gadi are provided in [Infrastructure usage and recommendations](#infrastructure-usage-and-recommendations) .
```

georgiesamaha commented on README.md:
````diff
@@ -123,25 +133,53 @@ The important features are:
 
+
+This pipeline has been tested for NCI Gadi HPC and the paramaters `--whoami` and `--gadi_account` will have to be included as discussed in the section [Infrastructure usage and recommendations](#infrastructure-usage-and-recommendations) .
+```
+nextflow run main.nf --input samples.csv \
+                     --ref /path/to/ref.fasta --dict /path/to/ref.dict \
+                     --ponvcf /path/to/pon \
+                     --common_biallelic_variants /path/to/common_biallelic_variants \
+                     -profile gadi \
+                     --whoami $(whoami) --gadi_account $PROJECT    
+```
````
Suggest moving this to [Infrastructure usage and recommendations](#infrastructure-usage-and-recommendations) to reduce repetition

georgiesamaha commented on README.md:
````diff
@@ -123,25 +133,53 @@ The important features are:
+nextflow run main.nf --input samples.csv \
+                     --ref /path/to/ref.fasta --dict /path/to/ref.dict \
+                     --ponvcf /path/to/pon \
+                     --common_biallelic_variants /path/to/common_biallelic_variants
 ```
-**Note**: Please use the command provided in the script `scripts/run_pipeline_on_gadi_script.sh` on NCI Gadi HPC.
 
 By default, this will generate `work` directory, `results` output directory and a `runInfo` run metrics directory inside the results directory. 
 
+
+This pipeline has been tested for NCI Gadi HPC and the paramaters `--whoami` and `--gadi_account` will have to be included as discussed in the section [Infrastructure usage and recommendations](#infrastructure-usage-and-recommendations) .
+```
+nextflow run main.nf --input samples.csv \
+                     --ref /path/to/ref.fasta --dict /path/to/ref.dict \
+                     --ponvcf /path/to/pon \
+                     --common_biallelic_variants /path/to/common_biallelic_variants \
+                     -profile gadi \
+                     --whoami $(whoami) --gadi_account $PROJECT    
+```
+
+
 To specify additional optional tool-specific parameters, see what flags are supported by running:
 
 ```
 nextflow run main.nf --help 
 ```
 
+A more advanced nextflow command with optional parameters, where you can define the name of the output folder and provide an integer value for the number of genomic-interval files to be generated looks like this:
+```
+nextflow run main.nf --input samples.csv \
+                     --ref /path/to/ref.fasta --dict /path/to/ref.dict \
+                     --ponvcf /path/to/pon \
+                     --common_biallelic_variants /path/to/common_biallelic_variants \
+                     -profile gadi \
+                     --whoami $(whoami) --gadi_account $PROJECT \
+                     --outDir name_output_folder \
+                     --number_of_intervals INT
+```
+
+
````
This doesnt adequately explain the use of parameters. This section should explain all parameters and when you might need to use them.

georgiesamaha commented on README.md:
```diff
@@ -123,25 +133,53 @@ The important features are:
 
+
+
 ### NCI Gadi HPC
```
Currently you have information required to execute workflow on Gadi split across various parts of the README. You need to commit to one part of the README that will contain all of this information. See comment above re: infrastructure usage recommendations section.

georgiesamaha commented on README.md:
```diff
@@ -162,19 +200,21 @@ You can then submit the script using the command:
 ### 5. Results 
 Once the pipeline is complete, you will find all outputs for each sample in the `results` directory. 
 
 ## Infrastructure usage and recommendations
```
See comments above about clearly capturing execution information for Gadi in one place

georgiesamaha commented on scripts/gatk4_selectvariants.pbs:
Have you tested this works?

nandan75 replied:
- Yes it works . 
- A script with parameters for test run is included at this time in `scripts/` folder. 
- I will delete the test script.

georgiesamaha commented on README.md:
```diff
@@ -63,38 +63,48 @@ When you run the pipeline, you will use the mandatory `--input` parameter to spe
-The user can create the Common biallelic variant resource files using our [pipeline](https://github.com/Sydney-Informatics-Hub/Somatic-ShortV?tab=readme-ov-file#0-optional-create-common-biallelic-variant-resources)
 
-### 4. Clone this repository 
+#### Create common biallelic variant resources for GetPileupSummaries 
+- A bash script `gatk4_selectvariants.pbs` is provided in the `scripts` folder which can read in your public resource VCF, and select common biallelic SNP variants (by default, those with an allele frequency of > 0.05).
+- This script can be directly run on the NCI Gadi HPC but it can also be modified to run on any other computational resource.
+- In the `gatk4_selectvariants.pbs` script, you can replace `<>` with your resource for resource=`<path/to/public_dataset.vcf.gz>` and output file resource_common=`</path/to/output_public_dataset_common_biallelic.vcf.gz>`. 
+- On NCI Gadi HPC you can adjust the compute resouces and submit your job using the command :`qsub gatk4_selectvariants.pbs`.
+- We will include this functionality as an optional module in the next version of the pipeline.
```
Why is this in dot points when other sections aren't? Please be more consistent with your formatting and try format things in a way that is easy to read for your user.

georgiesamaha reviewed (changes requested):
Still changes that need to be made. Things that need resolving: 
* Lack of details in README re: use of parameters 
* Inconsistent formatting of various sections 
* A single place in the README for instrucitons for running on Gadi. 

I suggest you take a step back and test your documentation to prepare all inputs and execute the pipeline. That will help you identify gaps and inconsistencies that make it hard to follow.

Commit:
Updated Readme.md  #5

Commit:
Updated Readme.md  #5

Commit:
Updated typos Readme.md  #5

Commit:
Tested create common allelicV, gatk version for makePON in Readme.md  #5

Commit:
gatk version updated gatk4_selectvariants.pbs  #5

georgiesamaha commented on README.md:
```diff
@@ -52,49 +53,48 @@ SAMPLE1,/data/Bams/sample1-N.bam,/data/Bams/sample1-T.bam
-#### A Panel of Normals (PoN) for Mutect2  
-The user can create a PoN file using our [pipeline](https://github.com/Sydney-Informatics-Hub/Somatic-ShortV?tab=readme-ov-file#4-create-pon-per-genomic-interval).
+**Common biallelic variant resources**  
+Common biallelic variant resources, such as databases that catalog known genetic variants (e.g., gnomAD) can be used to filter out germline variants and avoid false positives. 
+- A bash script `gatk4_selectvariants.pbs` is provided in the `scripts` folder which can read in your public resource VCF, and select common biallelic SNP variants (by default, those with an allele frequency of > 0.05). This script can be directly run on the NCI Gadi HPC. You can fill in the values for the required variables in this script and submit your job using the command :`qsub gatk4_select ...
+- We will include this functionality as an optional module in the next version of the pipeline.
```
```suggestion
Common biallelic variants databases (e.g. gnomAD) can be used to filter false positive germline variants from your somatic variant dataset. We have provided a script `scripts/gatk4_selectvariants.pbs` for this purpose. It executes[ GATK's SelectVariants](https://gatk.broadinstitute.org/hc/en-us/articles/13832694334235-SelectVariants) tool. This will be included as an optional functionality in this pipeline in an upcoming update.  
```

nandan75 replied:
ok

georgiesamaha commented on README.md:
````diff
@@ -52,49 +53,48 @@ SAMPLE1,/data/Bams/sample1-N.bam,/data/Bams/sample1-T.bam
 
-#### Common biallelic variant resources for GetPileupSummaries 
-The user can create the Common biallelic variant resource files using our [pipeline](https://github.com/Sydney-Informatics-Hub/Somatic-ShortV?tab=readme-ov-file#0-optional-create-common-biallelic-variant-resources)
+When you run the pipeline, you will use the mandatory `--common_biallelic_variants` parameter to specify the location and name of the reference.fasta file:
+```
````
Do you mean vcf?

nandan75 replied:
Yes I Meant vcf . Thanks

georgiesamaha commented on README.md:
````diff
@@ -138,11 +139,55 @@ To specify additional optional tool-specific parameters, see what flags are supp
 ```
 nextflow run main.nf --help 
 ```
+A more advanced nextflow command with optional parameters, where you can define the name of the output folder and provide an integer value for the number of genomic-interval files to be generated looks like this:
+```
+nextflow run main.nf --input samples.csv \
+                     --ref /path/to/ref.fasta --dict /path/to/ref.dict \
+                     --ponvcf /path/to/pon \
+                     --common_biallelic_variants /path/to/common_biallelic_variants \
+                     --outDir name_output_folder \
+                     --number_of_intervals INT
+```
````
Remove this. I don't think we should be saying this is more advanced, your dot points below explain what each of the mandatory and optional parameters do.

georgiesamaha commented on README.md:
```diff
@@ -138,11 +139,55 @@ To specify additional optional tool-specific parameters, see what flags are supp
+
+**Optional parameters**  
+* `--outDir`                        Name of the results directory                                         (**default:** `results`)  
+* `--number_of_intervals`           Define a specific number genomic-intervals for parallelisation        (**default:** `number of intervals are calculated using genome size`)  
```
```suggestion
* `--number_of_intervals`           Define a specific number genomic-intervals for parallelisation        (**default:** automatically calculated using genome size)  
```

nandan75 replied:
ok

georgiesamaha commented on README.md:
```diff
@@ -138,11 +139,55 @@ To specify additional optional tool-specific parameters, see what flags are supp
+* `--number_of_intervals`           Define a specific number genomic-intervals for parallelisation        (**default:** `number of intervals are calculated using genome size`)  
+
 
 If for any reason your workflow fails, you are able to resume the workflow from the last successful process with `-resume`. 
```
```suggestion
If for any reason your workflow fails, you are able to resume the workflow from the last successful process with `[-resume](https://www.nextflow.io/docs/latest/getstarted.html#modify-and-resume)`. 
```

georgiesamaha commented on README.md:
```diff
@@ -138,11 +139,55 @@ To specify additional optional tool-specific parameters, see what flags are supp
+This pipeline has been optimised for NCI Gadi HPC, instructions for executing on Gadi are provided in [Infrastructure usage and recommendations](#infrastructure-usage-and-recommendations)
+
+### 5. Results 
+Once the pipeline is complete, you will find all outputs for each sample in the `results` directory. 
```
Going to create a separate issue for this so it can be addressed in a new pull request, we need to explain what the outputs are and how they are organised.

georgiesamaha commented on README.md:
```diff
@@ -151,6 +196,10 @@ module purge
+
+
+
 To run this workflow on NCI Gadi HPC, you can excecute the script `scripts/run_pipeline_on_gadi_script.sh` by first entering the following details in the PBS header of the script:
 - project code
 - Resource-related details  
```
Also going to create another issue for this to be addressed in the benchmarking PR. The mem and ncpu resources consumed by the head nextflow job submitted using your pbs script will be standard for all runs and we should just provide them.

georgiesamaha commented on README.md:
Have added suggestions for a couple of things that aren't very clear. Will create new issues to be addressed in your benchmarking work.

georgiesamaha commented on config/gadi.config:
Why have all processes been given 8cpus instead of 1cpu in this pull request? GATK notoriously does not multithread so is it capable of consuming 8 cpus for every single one of these processes?

nandan75 replied:
Only relavant step is mutect2 which using intervals and hence multi-theads. I will conform on the actual resources when working on benchmarking - separate branch

georgiesamaha commented on nextflow.config:
Same comment as per your `gadi.config`. Can each of these GATK tools use 8 CPUs and is this relevant to the intervals work?

georgiesamaha replied:
80GB is also very large, was this necessary?

nandan75 replied:
Multi-threadign was only for the intervals work (so mutect2 only I guess - but will check).. I will look into this and other resource related issues when benchmarking.

georgiesamaha commented on scripts/gatk4_selectvariants_test.pbs:
```diff
@@ -0,0 +1,56 @@
+#
+#########################################################
+
+#PBS -P er01
+#PBS -N selectvariants
+#PBS -l walltime=04:00:00,ncpus=48,mem=192GB,wd
+#PBS -q normal
+#PBS -W umask=022
+#PBS -l storage=scratch/er01+gdata/er01
+#PBS -o ./Logs/selectvariants.o
+#PBS -e ./Logs/selectvariants.e
```
Please convert to template, remove project codes.

georgiesamaha commented on scripts/gatk4_selectvariants_test.pbs:
```diff
@@ -0,0 +1,56 @@
+#PBS -q normal
+#PBS -W umask=022
+#PBS -l storage=scratch/er01+gdata/er01
+#PBS -o ./Logs/selectvariants.o
```
Has this `./Logs` directory already been made? It will not be automatically made for users upon execution so you need to place the logs somewhere else.

nandan75 replied:
ok

georgiesamaha reviewed (changes requested):
Still a few changes to be made, I have made suggestions you can either choose to accept or reject and do yourself instead. I will make new issues to address issues with documentation and resources not relevant to this PR.

Commit:
Updated README.md deleted test script for selectvariant issue #5

georgiesamaha commented:
Closes #5.
